addons/sale/models/payment_transaction.py

```diff
@@ -92,33 +93,36 @@ def _log_message_on_linked_documents(self, message):
 
     def _reconcile_after_done(self):
         """ Override of payment to automatically confirm quotations and generate invoices. """
-        sales_orders = self.mapped('sale_order_ids').filtered(lambda so: so.state in ('draft', 'sent'))
-        for tx in self:
-            tx._check_amount_and_confirm_order()
+        confirmed_orders = self._check_amount_and_confirm_order()
         # send order confirmation mail
-        sales_orders._send_order_confirmation_mail()
-        # invoice the sale orders if needed
-        self._invoice_sale_orders()
-        res = super()._reconcile_after_done()
-        if self.env['ir.config_parameter'].sudo().get_param('sale.automatic_invoice') and any(so.state in ('sale', 'done') for so in self.sale_order_ids):
-            self.filtered(lambda t: t.sale_order_ids.filtered(lambda so: so.state in ('sale', 'done')))._send_invoice()
-        return res
+        confirmed_orders._send_order_confirmation_mail()
+
+        # invoice the sale orders if needed and send it
+        if self.env['ir.config_parameter'].sudo().get_param('sale.automatic_invoice'):
+            self._invoice_sale_orders()
```

Comment:
Little comment to explain why we do not reuse `confirmed_orders` here (bc of partial payments & co).